components/playlist/playlist_tab_helper.cc

```diff
@@ -0,0 +1,66 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/playlist/playlist_tab_helper.h"
+
+#include "base/feature_list.h"
+#include "brave/components/playlist/features.h"
+
+namespace playlist {
+
+PlaylistTabHelper::PlaylistTabHelper(content::WebContents* contents)
+    : content::WebContentsObserver(contents),
+      content::WebContentsUserData<PlaylistTabHelper>(*contents) {}
+
+PlaylistTabHelper::~PlaylistTabHelper() = default;
+
+// static
+void PlaylistTabHelper::HideMediaSrcAPIFromContents(
+    content::WebContents* contents) {
+  if (!base::FeatureList::IsEnabled(features::kPlaylist))
+    return;
+
+  DCHECK(contents && contents->GetMainFrame());
+
+  // This script is from
+  // https://github.com/brave/brave-ios/blob/development/Client/Frontend/UserContent/UserScripts/PlaylistSwizzler.js
+  static const std::u16string kScriptToHideMediaSourceAPI =
+      uR"-(
+    (function() {
+      // Stub out the MediaSource API so video players do not attempt to use `blob` for streaming
+      if (window.MediaSource || window.WebKitMediaSource || window.HTMLMediaElement && HTMLMediaElement.prototype.webkitSourceAddId) {
+        window.MediaSource = null;
+        window.WebKitMediaSource = null;
+        delete window.MediaSource;
+        delete window.WebKitMediaSource;
+      }
+    })();
+    )-";
+
+  const auto execute_script_callback = [](std::u16string script,
+                                          content::RenderFrameHost* frame) {
+#if BUILDFLAG(IS_ANDROID)
+    frame->ExecuteJavaScript(kScriptToHideMediaSourceAPI, base::NullCallback());
+#else
+    // In order to hide js API from main world, use testing api temporarily.
+    // TODO(sko) On Dekstop, it's not allowed to inject script like this.
+    // should we find another way?
+    frame->ExecuteJavaScriptForTests(kScriptToHideMediaSourceAPI,
+                                     base::NullCallback());
+#endif
+  };
+
+  contents->GetMainFrame()->ForEachRenderFrameHost(base::BindRepeating(
+      execute_script_callback, kScriptToHideMediaSourceAPI));
+}
+
+void PlaylistTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  HideMediaSrcAPIFromContents(web_contents());
```

Comment:
this is now called for all navigations, including non-main frame ones and same document ones.
this is called for all tabs even if playlist feature wasn't used at all (just enabled).

why?

Reply:
If we don't hide MediaSource API ahead of media detection script, we will get `blob:` source URL, which we can't download from.

Reply:
my point is that it will happen in all pages, even if user doesn't care about playlists in any way. Going this path we just can disable these endpoints in blink, but I don't think this is what we want.

Reply:
> Going this path we just can disable these endpoints in blink, but I don't think this is what we want.

Aha. I see where you're coming from now. We're going to support a global preference or content setting for this behavior. But for now we don't have any specific plan.